prov/efa/src/rdm/rxr_msg.c

```diff
@@ -233,6 +207,55 @@ ssize_t rxr_msg_generic_send(struct fid_ep *ep, const struct fi_msg *msg,
 		goto out;
 	}
 
+	if (peer->is_local && rxr_ep->use_shm_for_tx) {
+		/*
+		 * AWS Neuron and SynapseAI are currently not supported by the SHM provider.
+		 */
+		if (msg->desc && msg->desc[0] && (efa_mr_is_neuron(msg->desc[0]) || efa_mr_is_synapseai(msg->desc[0]))) {
+			EFA_WARN(FI_LOG_CQ,
+			"Hmem iface: %s is currently not supported by the SHM provider\n",
+			fi_tostr(&((struct efa_mr *)msg->desc[0])->peer.iface, FI_TYPE_HMEM_IFACE));
+			return -FI_EINVAL;
+		}
+
+		tmp_desc = msg->desc;
+		tmp_addr = msg->addr;
+		shm_msg = (struct fi_msg *) msg;
+		shm_msg->addr = peer->shm_fiaddr;
+		if (msg->desc) {
+			memcpy(shm_desc, msg->desc, msg->iov_count * sizeof(void *));
+			rxr_convert_desc_for_shm(msg->iov_count, shm_desc);
+			shm_msg->desc = shm_desc;
+		} else {
+			shm_msg->desc = NULL;
+		}
+		if (op == ofi_op_msg) {
```

Comment:
This could reduce code duplication, but I think it could cause more instructions executed. But if you make shm call in a higher level, all the changes @wenduwan  made in EFA for gdrcopy hacks would be copied and pasted everywhere.

Reply:
Fixed it in the newest revision.